espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/Http/Router.java

```diff
@@ -98,32 +119,78 @@ public BaseResponse handle(IHTTPSession session, Map<String, Object> params) {
                 String[] uriTokens = uri.split("/");
                 for (Map.Entry<Integer, String> wildcardIndexEntry : wildcardIndices.entrySet()) {
                     int wildcardIndex = wildcardIndexEntry.getKey();
-                    params.put(wildcardIndexEntry.getValue(), uriTokens[wildcardIndex]);
+                    uriParams.put(wildcardIndexEntry.getValue(), uriTokens[wildcardIndex]);
                 }
                 handler = routerMap.get(method).get(entry.getKey());
+                paramClass = paramClassMap.get(method).get(entry.getKey());
                 break;
             }
         }
 
-        BaseResponse res = handler.handle(session, params);
-        System.out.println("Finished processing " + method + " request for '" + uri + "'");
-        return res;
+        if (handler == null) {
+            return new ErrorResponse(AppiumStatus.UNKNOWN_COMMAND);
+        }
+
+        // Parse it to an Appium param
+        String postJson = parseBody(session);
+        AppiumParams appiumParams;
+        // TODO: Need to find a way to specify 'required' fields and throw exception when not provided
+        if (postJson == null) {
+            appiumParams = new AppiumParams();
+        } else {
+            appiumParams = (AppiumParams) paramClass.cast((new Gson()).fromJson(postJson, paramClass));
+        }
+        appiumParams.setSessionId(uriParams.get("sessionId"));
+        appiumParams.setElementId(uriParams.get("elementId"));
+
+        // Validate the sessionId
+        if (appiumParams.getSessionId() != null && !appiumParams.getSessionId().equals(Session.getGlobalSessionId())) {
+            return new ErrorResponse(AppiumStatus.BAD_PARAMETERS_ERROR, "Invalid session ID " + appiumParams.getSessionId());
+        }
+
+        // Validate the elementId
+        // TODO: Add a method to check if element is stale
+        if (appiumParams.getElementId() != null && !Element.exists(appiumParams.getElementId())) {
```

Comment:
In that case, we should have the handlers extend some abstract 'ElementHandler' object which does the checking.